### PR TITLE
GH-77 Cuckoo Cycle evidence added to headers

### LINKED ISSUE #20: Mine block using Cuckoo Cycle PoW
Depends on 

- [ ] #5

### LINKED ISSUE #67: Mine a block
Tracking ticket for mining.

- [ ] [Mine using SHA256](10).
- [ ] [Mine using Cuckoo Cycle](20).

### LINKED ISSUE #77: Add Cuckoo Cycle solution to the headers
Need to add the 42 ints to the header for ease of verification. See #5

### LINKED ISSUE #120: .gitignore missing exclusions
`git status` after build gives:
```
On branch master
Your branch is up-to-date with 'origin/master'.

Untracked files:
  (use "git add <file>..." to include in what will be committed)

	apps/aecore/c_src/cuckoo/
	apps/aecore/priv/

nothing added to commit but untracked files present (use "git add" to track)
```
We should exclude those paths

### PR DESCRIPTION
**Pull request by akorosmezey**
Refactored PoW generation so that SHA-256 and Cuckoo are
interchangeable.

In order to speed up unit tests, the Cuckoo NIF is compiled with
options that result in a 256-times smaller memory footprint.

Closes #77
Closes #20
Fixes #120
Closes #67

### CODE REVIEW

File: apps/aecore/src/aec_mining.erl [hunk 1]
```diff
@@ -8,7 +8,10 @@
 -include("blocks.hrl").
 -include("txs.hrl").
 
--define(DEFAULT_MINE_ATTEMPTS_COUNT, 10).
+
+%% Cuckoo Cycle has a solution probability of 2.2% so we are
+%% likely to succeed in 100 steps (~90%).
+-define(DEFAULT_MINE_ATTEMPTS_COUNT, 100).
```
Comment (wagerlabs) [inline apps/aecore/src/aec_mining.erl]: @sennui Consider making this a governance parameter.

**Comment by lucafavatella** [comment]
@akorosmezey It appears unit tests are not deterministically passing?

I see that one of the two CI jobs running the unit tests on commit 1bfb0dbb40033434f5584ac192a6749e7448de98 failed - see excerpt from https://travis-ci.org/aeternity/epoch/jobs/274697520 :
```
Failures:
  1) aec_mining_tests:mine_block_test_/0: Too few blocks mined in time increases new block's difficulty (PoW module aec_pow_cuckoo)
     Failure/Error: ?assertEqual(123, Block # block . nonce)
       expected: 123
            got: 188
     %% /home/travis/build/aeternity/epoch/_build/test/lib/aecore/test/aec_mining_tests.erl:137:in `aec_mining_tests:-mine_block_test_/0-fun-24-/1`
     Output: 
     Output: 
```

File: apps/aecore/c_src/Makefile [hunk 2]
```diff
@@ -100,6 +99,10 @@ help: ## This documentation
 
 all: deps $(LIB_SO) ## Build all
 
+# Use 256-times smaller memory footprint for nodes/edges in order to speed up unit tests
+test: all
+      CXXFLAGS += -DEDGEBITS=19
```
Comment (lucafavatella) [inline apps/aecore/c_src/Makefile]: A comment on the choice of number `19` would help - both for readability and for clarifying whether overflow is possible if e.g. an argument from the Erlang (test) invocation is changed.

**Comment by lucafavatella** [comment]
@akorosmezey Could you please add a comment to fields `#block.difficulty` and `#header.difficulty` in the record definitions of header and block, so to ease their usage (e.g. sum, comparison)? Also their default value (currently `0` - maybe to be removed?) and type (currently `non_neg_integer()`) may be revisited.

I see that `aec_pow:recalculate_difficulty/3` is the function that makes fields `#block.difficulty` and `#header.difficulty` be of type `aec_pow:sci_int()`. See also spec of `aec_pow_cuckoo:generate/3`.
At the moment `#block.difficulty` and `#header.difficulty` are defined with default value `0` and type `non_neg_integer()`.

File: apps/aecore/src/aec_pow.erl [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+pick_nonce() ->
+    rand:uniform(?NONCE_RANGE).
+
+%%------------------------------------------------------------------------------
+%% Adjust difficulty so that generation of new blocks proceeds at the expected pace
+%%------------------------------------------------------------------------------
+-spec recalculate_difficulty(sci_int(), integer(), integer()) -> sci_int().
+recalculate_difficulty(Difficulty, Expected, Actual) ->
+    DiffInt = scientific_to_integer(Difficulty),
+    integer_to_scientific(max(1, DiffInt * Expected div Actual)).
```
Comment (lucafavatella) [inline apps/aecore/src/aec_pow.erl]: I understand precedence of operators `*` and `div` are relevant in this computation. Parentheses would help.

**Comment by akorosmezey** [comment]
@lucafavatella Yes, I noticed that finding the solution is not deterministic. This happens because Cuckoo may drop a possible solution in the trimming phase. I rewrite the tests so that, instead of matching an expected nonce, we verify the result and match the result of verification.

**Comment by cytadela8** [comment]
This closes #120

File: apps/aecore/include/pow.hrl [hunk 4]
```diff
@@ -0,0 +1,4 @@
+-define(HIGHEST_BITCOIN_TARGET_SCI, 16#1d00ffff).
+-define(HIGHEST_BITCOIN_TARGET_INT, 16#00000000FFFF0000000000000000000000000000000000000000000000000000).
```
[the remark this answers is not in the document]
Comment (akorosmezey) [reply apps/aecore/include/pow.hrl]: The standard definition of bitcoin uses it. I am a bit uncertain what to use. We can stay compatible but, on the other hand, we need much lighter difficulties.

**Pull request merged**